bridge/sam/pool.go

```diff
@@ -0,0 +1,399 @@
+
+	for _, addr := range arr1 {
+		if !arr2Map[addr] {
+			diff = append(diff, addr)
+		}
+	}
+
+	return diff
+}
+
+func NewPool(validators []types.Address, threshold uint64) Pool {
+	return &pool{
+		changeValidatorsLock: sync.RWMutex{},
+		validators:           validators,
+		threshold:            threshold,
+		messageMap:           sync.Map{},
+		consumedMap:          sync.Map{},
+		readyMap:             sync.Map{},
+		messageSignatures:    newMessageSignaturesStore(),
+	}
+}
+
+// Add adds new message with the signature to pool
+func (p *pool) AddMessage(message *Message) {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	if p.hasConsumed(message.Hash) {
+		// we do no longer put the signature if the message has been consumed
+		return
+	}
+
+	p.setMessageBody(message.Hash, message.Body)
+	p.tryToPromote(message.Hash)
+}
+
+// Add adds new message with the signature to pool
+func (p *pool) AddSignature(signature *MessageSignature) {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	if p.hasConsumed(signature.Hash) {
+		// we do no longer put the signature if the message has been consumed
+		return
+	}
+
+	p.messageSignatures.Put(signature)
+	p.tryToPromote(signature.Hash)
+}
+
+// Consume sets the consumed flag and delete the message from pool
+func (p *pool) Consume(hash types.Hash) {
+	p.consumedMap.Store(hash, true)
+
+	p.messageSignatures.RemoveMessage(hash)
+	p.messageMap.Delete(hash)
+	p.readyMap.Delete(hash)
+}
+
+// knows returns the flag indicating the message is known
+func (p *pool) knows(hash types.Hash) bool {
+	_, ok := p.messageMap.Load(hash)
+	if !ok {
+		return false
+	}
+
+	return ok
+}
+
+// consumed returns the flag indicating the message is consumed
+func (p *pool) hasConsumed(hash types.Hash) bool {
+	raw, ok := p.consumedMap.Load(hash)
+	if !ok {
+		return false
+	}
+
+	consumed, ok := raw.(bool)
+
+	return ok && consumed
+}
+
+// GetReadyMessages returns the messages with enough signatures
+func (p *pool) GetReadyMessages() []ReadyMessage {
+	p.changeValidatorsLock.RLock()
+	defer p.changeValidatorsLock.RUnlock()
+
+	res := make([]ReadyMessage, 0)
+
+	p.readyMap.Range(func(key, value interface{}) bool {
+		hash, _ := key.(types.Hash)
+
+		if ready, _ := value.(bool); !ready {
+			return true
+		}
+
+		body := p.getMessageBody(hash)
+		if body == nil {
+			return true
+		}
+
+		signatures := p.messageSignatures.GetSignatures(hash)
+		res = append(res, ReadyMessage{
+			Body:       body,
+			Hash:       hash,
+			Signatures: signatures,
+		})
+
+		return true
+	})
+
+	return res
+}
+
+// UpdateValidatorSet update validators and threshold
+// This process blocks other processes because messages would lose the signatures
+func (p *pool) UpdateValidatorSet(validators []types.Address, threshold uint64) {
+	p.changeValidatorsLock.Lock()
+	defer p.changeValidatorsLock.Unlock()
+
+	oldValidators := p.validators
+	oldThreshold := p.threshold //nolint
```

Comment:
Why the `//nolint`?

Reply:
Actually it's necessary. It throws error because this declaration is far from for-loop which use this variable. But need to set here

Reply:
Aha, got it.

What's the linter that fails? if-short?

Reply:
ifshort

Reply:
It's some bug with the ifshort linter itself, you can then change the `nolint` to `nolint:ifshort` 👍

Reply:
Added 95405e5209abfed5ccbad58db15f2550f6d04bf5